### DISCUSSION TITLE
Is TypeAdapter thread safe?

### DISCUSSION THREAD

**Discussion by EldarSehayekZenity** [pydantic/pydantic#10762] [Question] [answered]
The docs for TypeAdapater recommend creating it once per adapted type due to some non triviial overhead.
Can I trust it to be thread-safe and indeed create a single instance per adapted type and import it wherever I need it? Should I make it thread-local instead to keep small number of TypeAdapter instances but avoid thread safety issues for them?

Would be glad to make the docs reflect the better the threading properties of TypeAdapter

**Comment by Viicos** [comment, marked as answer]
I'm assuming you'll be using type adapters to validate/serialize values? In that case yes, it should be thread safe. There's no mutation of any internal stuff happening when you call the `validate_*`/`dump_*` methods. Just make sure that the `TypeAdapter` instance is fully built (it will be if accessing the `core_schema` property succeeds).
